### PR TITLE
chore(service-proxy): global refactoring & bug resolution

### PR DESCRIPTION
**Pull request by sidux**
This a rework of service proxy + a port from the use case library.

- [x] Proxy generation bugs resolution (multiple annotations, multiple annotated methods, private props issues)
- [x] Make library easily extensible (a minimum of two classes only needed to add a new feature, that can also be loaded from  the parent project) for example a `@Segment` annotation can easily be loaded from the Sdz project
- [x] Handle `@Transaction`
- [x] Handle `@Event` with the possibility to send multiple events by using multiple annotations and the select an event handler at annotation level, for example using a Segment event handler the messenger, or the standard symfony.
- [x] Handle `@Security` with the possibility the have an AND by using multiple annotations
- [x] Add the possibility to select a provider/handler at annotation level, to resolve the current `@Cache` limitation, which makes it possible to have multiple providers in one class.
- [x] Fix ci
- [x] Add cs fixer
- [x] Add phpstan
- [x] Added symfony support as a bridge instead of using a different lib

### CODE REVIEW

File: .github/workflows/analysis.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                    -   name: Composer validate
+                        run: composer validate --ansi --strict
+
+                    -   name: PHPStan
+                        run: composer phpstan
+
+                    -   name: PHPCsFixer
+                        run: composer check-cs
+
+                    -   name: PHPUnit
```
Comment (NicolasMugnier) [inline .github/workflows/analysis.yaml:26]:
```suggestion
                    -   name: PHPCoverage
```

File: README.md [hunk 2]
```diff
@@ -30,123 +30,130 @@ use OpenClassrooms\ServiceProxy\ServiceProxy;
...
-use OpenClassrooms\ServiceProxy\Helpers\ServiceProxyHelper;
+Interceptors are used as decorators to react to the method execution. 
+for example using `@Cache` annotation, is a condition to enable the cache interceptor.
 
-$serviceProxyFactory = $this->getServiceProxyFactory();
-$proxy = $serviceProxyFactory->createProxy(new Class());
+There is two types of interceptors:
+#### Prefix interceptors :
+Interceptors that are called before the method execution, they must implement `OpenClassrooms\ServiceProxy\Interceptor\Contract\PrefixInterceptor`
+Two methods are called:
```
Comment (NicolasMugnier) [inline README.md:44]:
```suggestion
Three methods are called:
```

File: README.md [hunk 3]
```diff
@@ -30,123 +30,130 @@ use OpenClassrooms\ServiceProxy\ServiceProxy;
...
+- `prefix` : called before the method execution. Should return instance of `OpenClassrooms\ServiceProxy\Interceptor\Response\Response`.
+- `supportsPrefix` : called to know if the interceptor should be called, for example in the case of the cache interceptor, it will check that the method has the `@Cache` annotation.
+- `getPrefixPriority` : called to know the priority of the interceptor, the higher the priority, the earlier the interceptor will be called.
 
-```
+#### Suffix interceptors :
+Interceptors that are called after the method execution, even if an exception is thrown, they must implement `OpenClassrooms\ServiceProxy\Interceptor\Contract\SuffixInterceptor`
+Two methods are called:
```
Comment (NicolasMugnier) [inline README.md:51]:
```suggestion
Three methods are called:
```

File: README.md [hunk 4]
```diff
@@ -30,123 +30,130 @@ use OpenClassrooms\ServiceProxy\ServiceProxy;
...
-```
+Handlers are used by interceptors to manage the infrastructure code.
+To be able to use built-in interceptors, you need to implement the built-in handlers contracts.
 
-##### Builder
-``` php
-use OpenClassrooms\ServiceProxy\Helpers\ServiceProxyHelper;
+- All handles need to implement `OpenClassrooms\ServiceProxy\Contract\AnnotationHandler`.
+- Each handler must have a unique name, you can use the `getName` method to return it.
+- Each handler must return if it's the default handler, you can use the `isDefault` method to return it.
```
Comment (NicolasMugnier) [inline README.md:76]:
```suggestion
- Each handler must return ? if it's the default handler, you can use the `isDefault` method to return it.
```
Must return what ? Maybe Have I missed something.

File: docs/Interceptor/event.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+}
+```
+
+The message can be send:
+
+- pre execute
+- post execute
+- on exception
+  or all of them.
```
Comment (NicolasMugnier) [inline docs/Interceptor/event.md:35]:
```suggestion
- or all of them.
```

File: src/Annotation/Event.php [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public function getName(): ?string
+    {
+        return $this->name;
+    }
+
+    public function hasMethod(string $method): bool
+    {
+        return \in_array($method, $this->methods, true);
```
Comment (NicolasMugnier) [inline src/Annotation/Event.php:65]: Maybe can we apply `\` on all PHP functions. What do you think ?

File: tests/Double/Stub/ParameterClassStub.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenClassrooms\ServiceProxy\Tests\Double\Stub;
+
+/**
+ * @author Romain Kuzniak <romain.kuzniak@openclassrooms.com>
```
Comment (NicolasMugnier) [inline tests/Double/Stub/ParameterClassStub.php:8]: Should we keep `@author` annotation ?

File: src/Interceptor/AbstractInterceptor.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+namespace OpenClassrooms\ServiceProxy\Interceptor;
+
+use OpenClassrooms\ServiceProxy\Annotation\Annotation;
+use OpenClassrooms\ServiceProxy\Contract\AnnotationHandler;
+use OpenClassrooms\ServiceProxy\Contract\Exception\DuplicatedDefaultHandler;
+use OpenClassrooms\ServiceProxy\Contract\Exception\DuplicatedHandler;
+use OpenClassrooms\ServiceProxy\Contract\Exception\HandlerNotFound;
+use OpenClassrooms\ServiceProxy\Contract\Exception\MissingDefaultHandler;
+
+abstract class AbstractInterceptor
```
Comment (NicolasMugnier) [inline src/Interceptor/AbstractInterceptor.php:14]: Here as methods `public function getPrefixPriority(): int` and `public function getSuffixPriority(): int` are implemented and all subclasses implements `(Prefix|Suffix)Interceptor`,  why don't we implements both `SuffixInterceptor, PrefixInterceptor` interfaces in this abstract class and add the following methods as abstract : 

- `public abstract function prefix(Instance $instance): Response;`
- `public abstract function supportsPrefix(Instance $instance): bool;`
- `public abstract function suffix(Instance $instance): Response;`
- `public abstract function supportsSuffix(Instance $instance): bool;`

Comment (NicolasMugnier) [reply src/Interceptor/AbstractInterceptor.php:14]: Ok I can see below that there is a subclass that doesn't implements both interfaces, so I'm wondering why methods `public function getPrefixPriority(): int` and  `public function getSuffixPriority(): int` are implemented in this class except for some factorisation. 

Finally we can have subclasses which implements one of `(Prefix|Suffix)Interceptor` or both or why not another Interceptor interface, so here those methods are part of two distinct interfaces, I think they should be implemented in the real implementation even if the code is duplicated.

In other words why a "common" abstract class implement some methods (which can be unused) from different interfaces without implement these interfaces ?

File: src/Interceptor/EventInterceptor.php [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            }
+        }
+
+        return new Response();
+    }
+
+    public function supportsPrefix(Instance $instance): bool
+    {
+        return $instance->getMethod()
+            ->hasAnnotation(Event::class);
```
Comment (NicolasMugnier) [inline src/Interceptor/EventInterceptor.php:85]:
```suggestion
        return $this->supportsSuffix($instance);
```

Comment (NicolasMugnier) [reply src/Interceptor/EventInterceptor.php:85]: or in reverse, the idea is to keep the consistency with other interceptors, what do you think ?

File: src/Interceptor/AbstractInterceptor.php (hunk 8 above)
Comment (sidux) [reply src/Interceptor/AbstractInterceptor.php:14]: I see the issue, I used this as a shortcut to have priority defaults into 0, but I think you have a point, and it's better to force each implementation think about it's priority anyway.

File: src/Annotation/Event.php (hunk 6 above)
Comment (sidux) [reply src/Annotation/Event.php:65]: Done, added `NativeFunctionInvocationFixer` rule
